sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala

```diff
@@ -146,14 +147,17 @@ case class CoalesceShufflePartitions(session: SparkSession) extends AQEShuffleRe
       Seq(collectShuffleStageInfos(r))
     case unary: UnaryExecNode => collectCoalesceGroups(unary.child)
     case union: UnionExec => union.children.flatMap(collectCoalesceGroups)
-    // If not all leaf nodes are exchange query stages, it's not safe to reduce the number of
-    // shuffle partitions, because we may break the assumption that all children of a spark plan
-    // have same number of output partitions.
     // Note that, `BroadcastQueryStageExec` is a valid case:
     // If a join has been optimized from shuffled join to broadcast join, then the one side is
     // `BroadcastQueryStageExec` and other side is `ShuffleQueryStageExec`. It can coalesce the
     // shuffle side as we do not expect broadcast exchange has same partition number.
-    case p if p.collectLeaves().forall(_.isInstanceOf[ExchangeQueryStageExec]) =>
+    case join: BroadcastHashJoinExec => join.children.flatMap(collectCoalesceGroups)
```

Comment:
We should also match `CartesianProductExec` here.

Reply:
> We should also match `CartesianProductExec` here.

No get why we need match `CartesianProductExec` here. For `CartesianProductExec`, the two sides are both required, so there is no need to process `collectCoalesceGroups` separately.

Reply:
assume the children of `CartesianProductExec` are GROUP BY, then we can coalesce the shuffles of the two sides independently.

Reply:
Got it.